torch/testing/_internal/distributed/_sharded_tensor/__init__.py

```diff
@@ -27,6 +27,10 @@ def init_pg(self, backend="nccl"):
             init_method=f"file://{self.file_name}",
         )
 
+        # set device for nccl pg for collectives
+        if backend == "nccl":
+            torch.cuda.set_device(self.rank)
```

Comment:
We should update the docs of `ShardedTensor` to clearly mention this expectation from the user. We could actually even do an all_gather of all default device ids across all ranks of ShardedTensor and throw an error if the default device is the same on multiple ranks.

Reply:
updated the PR with the doc, quick question on the default device id, are you referring to that we can all_gather the `torch.cuda.current_device()` results from all rank and check if there's overlap?

Reply:
Let's add this in a follow up PR, I am actually a bit concerning on adding this, mainly because this means for every collective we run in nccl, we will need to run an additional all_gather just to check the default device overlap or not, which might potentially impact the perf?

Reply:
Created issue to track:
https://github.com/pytorch/pytorch/issues/70054
https://github.com/pytorch/pytorch/issues/70053